semmc-synthesis/src/SemMC/Synthesis/Cegis.hs

```diff
@@ -55,21 +55,22 @@ data CegisResult sym arch = CegisUnmatchable [ConcreteTest sym arch]
                           -- of the candidate instructions given, has the same
                           -- behavior as the target formula.
 
-cegis' :: forall arch t solver fs.
+cegis' :: forall arch solver t st fs.
           (A.Architecture arch, A.ArchRepr arch, A.Architecture (T.TemplatedArch arch)
           , WPO.OnlineSolver solver
-          , CB.IsSymInterface (CBO.OnlineBackend t solver fs)
+          , CB.IsSymInterface (WE.ExprBuilder t st fs)
           , ?memOpts :: LLVM.MemOptions
           )
-       => [T.TemplatedInstructionFormula (CBO.OnlineBackend t solver fs) arch]
+       => CBO.OnlineBackend solver t st fs
+       -> [T.TemplatedInstructionFormula (WE.ExprBuilder t st fs) arch]
        -- ^ The trial instructions.
-       -> Formula (CBO.OnlineBackend t solver fs) arch
+       -> Formula (WE.ExprBuilder t st fs) arch
        -- ^ A formula representing the sequence of trial instructions.
-       -> Cegis (CBO.OnlineBackend t solver fs) arch (CegisResult (CBO.OnlineBackend t solver fs) arch)
-cegis' trial trialFormula = do
+       -> Cegis (WE.ExprBuilder t st fs) arch (CegisResult (WE.ExprBuilder t st fs) arch)
+cegis' bak trial trialFormula = do
   sym <- askSym
```

Comment:
Could this use `askBackend` instead? That way, you wouldn't need to change the type of `cegis'` to add an explicit `bak` argument. (There are various other functions that now take an additional `bak` argument for which the same question applies.)

Reply:
I don't think so: the `askBackend` gives a `SomeBackend`, which existentially closes over the type. In some places (including here, I'm pretty sure) we need to actually know that we have an `OnlineBackend`. There are other ways, maybe, to achieve this, but this seemed like the easiest to me.